rfc/20241003-temporary-files.md

```diff
@@ -0,0 +1,295 @@
+# Explicit mechanism for working with Temporary Files
+
+Issues: This is indirectly motivated by both [#1086](https://github.com/opentofu/opentofu/issues/1086) and [#1942](https://github.com/opentofu/opentofu/issues/1942), but also represents a distinct feature in its own right.
+
+Although ideally an OpenTofu configuration would do all of its work in local RAM and remote APIs, there are some situations where the most appropriate technique is to write a temporary file to disk and then read that file back into memory in some other location.
+
+For example, when preparing large files to be uploaded into blob storage systems like Amazon S3 it's common to want to prepare the file using some other provider and then ask the `hashicorp/aws` provider to read that file and upload it into an Amazon S3 bucket (or equivalent for other blob storage services).
+
+OpenTofu has no built-in features for managing temporary files, and so informal patterns have emerged. The two most prominent approaches are:
+
+- Use `path.cwd` to generate a path referring to a file in the current working directory, and then pass that absolute path to all of the providers that need to write or read the file in question.
+
+    This keeps all of the generated files together in one place, but requires vigilance in managing what is effectively a shared namespace of filenames. The most robust form of this pattern involves the root module deciding all of the paths and then passing them as strings to other modules, so that the root module can act as the central coordinator of its own namespace.
+
+- Use `path.module` to generate a path relative to a specific module, and then have that module return its generated path as an output value so that other parts of the configuration can read the generated file.
+
+    This is largely the same as the first point but has the minor benefit of allowing each module to manage its own filesystem namespace. However, it's particularly problematic when a module is called using `for_each` or `count` because then multiple instances of the module all share a single directory, and will clobber each other's written files unless specialc care is taken.
```

Comment:
Typo: `specialc`.

I've got https://github.com/opentofu/opentofu/issues/1086 assigned to me to remind me to update the documentation with this fact / warning.